README.md
# HaTs Documentation Website

## Demo

View the site live at at [hats.dev](https://hats.dev)

## Description

HaTs (short for _Happy TypeScript_) is a project scaffolding CLI that speeds up
the process of creating TS libraries.

This repo contains the source code for the HaTs documentation site.

Stack:

- **Framework**: [Gatsby](https://www.gatsbyjs.com/)
- **Theme**: [smooth-doc](https://smooth-doc.com/)
- **Content**: [MDX](https://github.com/mdx-js/mdx)
- **Styling**: [@xstyled/styled-components](https://xstyled.dev/)
- **Hosting**: [Firebase](https://firebase.google.com)

## Author

- [Kamar Mack](https://github.com/kamarmack)

## License

BSD-3-Clause


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v1.1.2](https://github.com/hats-dev/hats-dev-web-app/compare/v1.1.1...v1.1.2)

- migrate mackbook.io to kamar.blog [`e38cac9`](https://github.com/hats-dev/hats-dev-web-app/commit/e38cac92f2ea043f61d3d78cf522af09d45efe35)

#### [v1.1.1](https://github.com/hats-dev/hats-dev-web-app/compare/v1.1.0...v1.1.1)

> 6 May 2022

- fix env variable read [`509ac67`](https://github.com/hats-dev/hats-dev-web-app/commit/509ac6713f428a95aae488f36cf4f09b42970f92)

#### [v1.1.0](https://github.com/hats-dev/hats-dev-web-app/compare/v1.0.4...v1.1.0)

> 6 May 2022

- add google analytics [`f3b21d4`](https://github.com/hats-dev/hats-dev-web-app/commit/f3b21d4892105be89802826ffcbf5415e018f988)

#### [v1.0.4](https://github.com/hats-dev/hats-dev-web-app/compare/v1.0.3...v1.0.4)

> 1 May 2022

- rename remote repo [`07bc15b`](https://github.com/hats-dev/hats-dev-web-app/commit/07bc15b74278be66a4a7bf84bd4945462e5c7a5f)
- improve readme copy [`5fc3233`](https://github.com/hats-dev/hats-dev-web-app/commit/5fc3233bf073128e5b7454b450ab0b8582e7a0aa)

#### [v1.0.3](https://github.com/hats-dev/hats-dev-web-app/compare/v1.0.2...v1.0.3)

> 30 April 2022

- change logo destination to docs [`c4dc202`](https://github.com/hats-dev/hats-dev-web-app/commit/c4dc202e51aef6bf1e95dc743b73d6a4bfc260d2)

#### [v1.0.2](https://github.com/hats-dev/hats-dev-web-app/compare/v1.0.1...v1.0.2)

> 30 April 2022

- patch trailing slash error [`9332e30`](https://github.com/hats-dev/hats-dev-web-app/commit/9332e305ba24cdb3ca4a293b1690f4cff14b7904)

#### [v1.0.1](https://github.com/hats-dev/hats-dev-web-app/compare/v1.0.0...v1.0.1)

> 30 April 2022

- patch destinations of internal links [`74a4969`](https://github.com/hats-dev/hats-dev-web-app/commit/74a49697a4da21750ebb5ece6311575dbeef49ea)

#### v1.0.0

> 30 April 2022

- set up gatsby project [`b47e459`](https://github.com/hats-dev/hats-dev-web-app/commit/b47e459bec73010df6415a511767cea0d8670de2)
- initial commit [`994a575`](https://github.com/hats-dev/hats-dev-web-app/commit/994a5750b71862ab821e86d463448bbec139979d)
- complete full draft of dev workflow copy edit [`62649a9`](https://github.com/hats-dev/hats-dev-web-app/commit/62649a9a2098766290c25558ba3af5082fdcdc9b)


src/smooth-doc/components/SideNav.js
import React from 'react';
import { useStaticQuery, graphql, Link } from 'gatsby';
import styled from '@xstyled/styled-components';
import { useLocation } from '@reach/router';

const SideNavQuery = graphql`
	query SideNavQueryCustom {
		allMdx(filter: { fields: { pageType: { eq: "doc" }, title: { ne: "" } } }) {
			edges {
				node {
					id
					fields {
						title
						pageType
						section
						order
						slug
					}
				}
			}
		}

		site {
			siteMetadata {
				sections
			}
		}
	}
`;

const createOrFindGroup = (name, groups) => {
	const existingGroup = groups.find((group) => group.name === name);
	if (existingGroup) return existingGroup;

	const group = { name, nodes: [] };
	groups.push(group);
	return group;
};

const DEFAULT_ORDER_VALUE = -9999;

const sortNodes = (a, b) => {
	const diff =
		a.fields.order !== DEFAULT_ORDER_VALUE &&
		b.fields.order !== DEFAULT_ORDER_VALUE
			? a.fields.order - b.fields.order
			: a.fields.order !== DEFAULT_ORDER_VALUE
			? -1
			: b.fields.order !== DEFAULT_ORDER_VALUE
			? 1
			: a.fields.title.localeCompare(b.fields.title);
	return diff === 0 ? 0 : diff > 0 ? 1 : -1;
};

const groupNodes = (nodes) =>
	nodes.reduce((groups, node) => {
		const group = createOrFindGroup(node.fields.section || 'Docs', groups);
		group.nodes.push(node);
		group.nodes.sort(sortNodes);
		return groups;
	}, []);

const Nav = styled.nav`
	padding: 4 3 5;
`;

const NavGroup = styled.div`
	margin-bottom: 4;
`;

const NavGroupTitle = styled.h4`
	font-size: 14;
	font-weight: 500;
	color: on-background-light;
	text-transform: uppercase;
	margin: 0 0 3 0;
`;

const NavGroupMenu = styled.ul`
	margin: 0;
	padding: 0;
	border-left: 1;
	border-left-color: layout-border;
`;

const NavGroupMenuItem = styled.li`
	list-style-type: none;
	margin: 2 0 0 -2px;
	padding: 0;
	font-size: 14;
	font-weight: 500;

	a {
		color: on-background-light;
		display: block;
		transition: fast;
		transition-property: color, border-color;
		padding: 2px 0 2px 2;
		border-left: 3;
		border-color: transparent;
		text-decoration: none;
		opacity: 0.85;

		&:hover {
			color: on-background;
			opacity: 1;
		}

		&[aria-current='page'] {
			font-weight: 600;
			border-color: primary;
			color: on-background;
			opacity: 1;
		}
	}
`;

const sortGroupsWithConfig = (section) => (a, b) => {
	const indexA = section.indexOf(a.name);
	const indexB = section.indexOf(b.name);
	const diff = indexA - indexB;
	return diff === 0 ? 0 : diff < 0 ? -1 : 1;
};

export function useSideNavState() {
	const data = useStaticQuery(SideNavQuery);
	return React.useMemo(() => {
		const navGroups = groupNodes(data.allMdx.edges.map((edge) => edge.node));
		navGroups.sort(sortGroupsWithConfig(data.site.siteMetadata.sections));
		return { navGroups };
	}, [data]);
}

export function useSideNavPrevNext({ navGroups }) {
	const { pathname } = useLocation();
	const nodes = navGroups.flatMap((group) => group.nodes);
	const nodeIndex = nodes.findIndex(
		(node) =>
			node.fields.slug.replace(/\/$/, '') === pathname.replace(/\/$/, ''),
	);
	return {
		prev: nodeIndex > -1 ? nodes[nodeIndex - 1] : null,
		next: nodeIndex > -1 ? nodes[nodeIndex + 1] : null,
	};
}

export function SideNav({ navGroups }) {
	return (
		<Nav>
			{navGroups.map((navGroup, index) => (
				<NavGroup key={index}>
					<NavGroupTitle>{navGroup.name}</NavGroupTitle>
					<NavGroupMenu>
						{navGroup.nodes.map((page) => (
							<NavGroupMenuItem key={page.id}>
								<Link to={page.fields.slug.replace(/\/$/, '')}>
									{page.fields.title}
								</Link>
							</NavGroupMenuItem>
						))}
					</NavGroupMenu>
				</NavGroup>
			))}
		</Nav>
	);
}


src/smooth-doc/components/DocLayout.js
import * as React from 'react';
import styled, {
	Box,
	css,
	up,
	down,
	th,
	useDown,
} from '@xstyled/styled-components';
import { useDialogState, Dialog, DialogDisclosure } from 'reakit/Dialog';
import { Portal } from 'reakit/Portal';
import { VscChevronUp } from 'react-icons/vsc';
import { RiPencilLine } from 'react-icons/ri';
import { ScreenContainer } from 'smooth-doc/src/components/ScreenContainer';
import { PageLayout } from 'smooth-doc/src/components/PageLayout';
import {
	SiblingNav,
	SiblingNavLink,
} from 'smooth-doc/src/components/SiblingNav';
import { Article } from 'smooth-doc/src/components/Article';
import { TableOfContents } from 'smooth-doc/src/components/TableOfContents';
import { SideNav, useSideNavState, useSideNavPrevNext } from './SideNav';

const SidebarDialog = styled.div`
	background-color: background-light-a50;
	backdrop-filter: blur(3px);
	position: fixed;
	top: 50;
	right: 0;
	bottom: 0;
	left: 0;
	z-index: 20;
	overflow: auto;
	transition: base;
	opacity: 0;
	transition: opacity 250ms ease-in-out, transform 250ms ease-in-out;
	transform: translate3d(0, 10vh, 0);

	&[data-enter] {
		opacity: 1;
		transform: translate3d(0, 0, 0);
	}

	&:focus {
		outline: none;
	}
`;

const Container = styled.div`
	z-index: 0;
	position: relative;

	${up(
		'lg',
		css`
			display: grid;
			grid-template-columns: 288px minmax(0, 1fr);
			grid-gap: ${th.space(5)};

			.sidebar-container {
				display: none;
			}
		`,
	)}

	${up(
		'xl',
		css`
			grid-template-columns: 288px minmax(0, 1fr) 288px;

			.sidebar-container {
				display: none;
			}
		`,
	)}
`;

const TocContainer = styled.div`
	${down(
		'xl',
		css`
			display: none;
		`,
	)}
`;

const SidebarSticky = styled.aside`
	position: sticky;
	top: ${th.px(50)};
	padding: 4 0;
	overflow-y: auto;
	height: calc(100vh - 50px);
	width: 288px;

	${down(
		'lg',
		css`
			display: none;
		`,
	)}
`;

const MenuButton = styled.button`
	appearance: none;
	border: 0;
	border-radius: 50%;
	width: 60;
	height: 60;
	position: fixed;
	right: ${th.size(8)};
	bottom: ${th.size(8)};
	z-index: 25;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	color: background;
	background-color: on-background;
	transition: base;
	transition-property: color;

	&:focus {
		color: background;
	}

	> svg {
		width: 24;
		height: 24;
		transition: transform 200ms ease-in-out;

		&:first-child {
			transform: translateX(-1px);
		}

		&:last-child {
			transform: translate(-1px) rotate(180deg);
		}
	}

	&[aria-expanded='true'] {
		> svg:first-child {
			transform: translate(-1px, 16px);
		}

		> svg:last-child {
			transform: translate(-1px, -16px) rotate(180deg);
		}
	}
`;

const Footer = styled.div`
	margin-top: 3em;
	text-align: center;
	font-weight: 300;
	color: on-background-light;
	${up(
		'md',
		css`
			margin-top: 8em;
		`,
	)}
`;

const AuthorLink = styled.aBox`
	font-size: 18;
	transition: fast;
	text-decoration: none;
	color: on-background-primary;

	&:hover {
		color: on-background-primary-dark;
	}
`;

function MobileSidebar({ children }) {
	const dialog = useDialogState({ animated: true });
	return (
		<>
			<Dialog {...dialog} as={SidebarDialog}>
				{children}
			</Dialog>
			<Portal>
				<DialogDisclosure {...dialog} as={MenuButton}>
					<VscChevronUp />
					<VscChevronUp />
				</DialogDisclosure>
			</Portal>
		</>
	);
}

function PrevNextLinks(props) {
	const { prev, next } = useSideNavPrevNext(props);
	if (!prev && !next) return null;
	return (
		<SiblingNav>
			{prev && (
				<SiblingNavLink
					type="previous"
					to={prev.fields.slug.replace(/\/$/, '')}
				>
					{prev.fields.title}
				</SiblingNavLink>
			)}
			{next && (
				<SiblingNavLink type="next" to={next.fields.slug.replace(/\/$/, '')}>
					{next.fields.title}
				</SiblingNavLink>
			)}
		</SiblingNav>
	);
}

export function DocLayout({ children, tableOfContents, editLink, ...props }) {
	const downLg = useDown('lg');
	const sideNav = useSideNavState();
	const year = new Date().getFullYear().toString();
	return (
		<PageLayout {...props}>
			<ScreenContainer px={0}>
				<Container>
					<SidebarSticky>
						<SideNav {...sideNav} />
					</SidebarSticky>
					<div className="sidebar-container">
						{downLg && (
							<MobileSidebar>
								<SideNav {...sideNav} />
							</MobileSidebar>
						)}
					</div>
					<Box pb={6} px={3}>
						<Article>
							{children}
							{editLink && (
								<Box
									mt={5}
									display="grid"
									gridTemplateColumns="min-content max-content"
									gridGap={1}
									alignItems="center"
									forwardedAs="a"
									href={editLink}
								>
									<RiPencilLine /> Edit this page on GitHub
								</Box>
							)}
							<PrevNextLinks {...sideNav} />
						</Article>
						<hr />
						<Footer textAlign={'center'}>
							<p>
								Happy TypeScript
								<br />©{year}
								&nbsp;
								<AuthorLink
									target="_blank"
									rel="noreferrer"
									href="https://kamar.blog"
								>
									👨🏾‍💻 Kamar Mack
								</AuthorLink>
							</p>
						</Footer>
					</Box>
					<TocContainer>
						<TableOfContents />
					</TocContainer>
				</Container>
			</ScreenContainer>
		</PageLayout>
	);
}


src/smooth-doc/components/AppHeader.js
import React from 'react';
import { graphql, useStaticQuery } from 'gatsby';
import styled, {
	Box,
	useColorMode,
	down,
	css,
} from '@xstyled/styled-components';
import { ScreenContainer } from 'smooth-doc/src/components/ScreenContainer';
import { DocSearch } from 'smooth-doc/src/components/DocSearch';
import { NavLink } from 'smooth-doc/src/components/Nav';
import { AppNav } from 'smooth-doc/src/components/AppNav';

const AppHeaderQuery = graphql`
	query AppHeaderCustom {
		logos: allFile(
			filter: {
				sourceInstanceName: { in: ["image", "default-image"] }
				name: { glob: "logo-nav*" }
			}
			sort: { fields: sourceInstanceName, order: DESC }
		) {
			nodes {
				name
				publicURL
			}
		}

		site {
			siteMetadata {
				title
				docSearch {
					apiKey
					indexName
				}
				navItems {
					title
					url
				}
			}
		}
	}
`;

const OuterHeader = styled.header`
	background-color: background;
	border-bottom-style: solid;
	border-bottom-width: base;
	border-bottom-color: layout-border;
	padding: 2 0;
	height: 50;

	${down(
		'sm',
		css`
			overflow-x: auto;
		`,
	)}
`;

const NavSkipLink = styled.a`
	text-decoration: none;
	left: -999px;
	position: absolute;
	top: auto;
	width: 1px;
	height: 1px;
	overflow: hidden;
	z-index: -999;
	color: primary;

	&:focus {
		background-color: background;
		border-radius: base;
		left: auto;
		top: auto;
		height: auto;
		width: max-content;
		overflow: auto;
		padding: 2;
		margin: 2;
		text-align: center;
		z-index: 999;
	}
`;

function useLogo(logos) {
	const [mode] = useColorMode();
	switch (mode) {
		case 'dark':
			return (
				logos.find((logo) => logo.name === 'logo-nav-dark') ||
				logos.find((logo) => logo.name === 'logo-nav') ||
				null
			);
		case 'light':
		default:
			return (
				logos.find((logo) => logo.name === 'logo-nav-light') ||
				logos.find((logo) => logo.name === 'logo-nav') ||
				null
			);
	}
}

export function AppHeader() {
	const data = useStaticQuery(AppHeaderQuery);
	const logo = useLogo(data.logos.nodes);

	return (
		<OuterHeader>
			<ScreenContainer>
				<NavSkipLink tabindex="0" href="#main">
					Skip to content
				</NavSkipLink>
				<Box row alignItems="center" flexWrap="nowrap" mx={-2}>
					<Box col px={2} display="flex">
						<NavLink
							row
							display="inline-flex"
							alignItems="center"
							flexWrap="nowrap"
							mx={-2}
							to="/docs/v1"
						>
							{logo ? (
								<Box
									col="auto"
									px={2}
									height={32}
									forwardedAs="img"
									src={logo.publicURL}
									alt={data.site.siteMetadata.title}
								/>
							) : (
								<Box
									col="auto"
									flex="0 1 auto"
									px={2}
									forwardedAs="h2"
									alignItems="center"
									fontSize={18}
									m={0}
								>
									{data.site.siteMetadata.title}
								</Box>
							)}
						</NavLink>
					</Box>
					{data.site.siteMetadata.docSearch ? (
						<Box col="auto" px={2} display={{ xs: 'none', md: 'block' }}>
							<DocSearch {...data.site.siteMetadata.docSearch} />
						</Box>
					) : null}
					<AppNav col="auto" px={2} />
				</Box>
			</ScreenContainer>
		</OuterHeader>
	);
}
